### PR TITLE
nixos/networkmanager: default firewallBackend to nftables, remove firewallBackend

### PR DESCRIPTION
**Pull request by SuperSandro2000**
###### Description of changes

see #161428

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/networkmanager.nix [hunk 1]
```diff
@@ -233,7 +233,8 @@ in {
 
       firewallBackend = mkOption {
         type = types.enum [ "iptables" "nftables" "none" ];
-        default = "iptables";
+        default = if config.networking.nftables.enable then "nftables" else "iptables";
+        defaultText = ''if config.networking.nftables.enable then "nftables" else "iptables"'';
```
Comment (jian-lin) [inline nixos/modules/services/networking/networkmanager.nix:237]: Please use `literalExpression`

**Review by jian-lin — DISMISSED** [review]

**Comment by mkg20001** [comment]
@flokli What's left (besides CI green) for this to be ready?

**Comment by mkg20001** [comment]
automatic detection simply checks if the nft executable exists and then uses nft right away, which i assume will always be the case in nixos since we patch the path in

meaning the effective default is nftables.

will it be more or less confusing if we set this explicitly?
can legacy iptables even still be enabled? otherwise we might want to get rid of the firewallBackend

````
static NMFirewallBackend
_firewall_backend_detect(void)
{
    if (g_file_test(NFT_PATH, G_FILE_TEST_IS_EXECUTABLE))
        return NM_FIREWALL_BACKEND_NFTABLES;
    if (g_file_test(IPTABLES_PATH, G_FILE_TEST_IS_EXECUTABLE))
        return NM_FIREWALL_BACKEND_IPTABLES;

    return NM_FIREWALL_BACKEND_NFTABLES;
}
```

**Comment by flokli** [comment]
I think people can still opt out of nftables by overwriting the `iptables` binary to the legacy one. At least the example in `networking.firewall.package` mentions it as an example. This is probably not tested at all, and not sure if it's worthwhile putting in that work still.

I'd probably just remove the `networking.networkmanager.firewallBackend`option entirely, and keep the autodetection code in NetworkManager itself.

**Comment by mkg20001** [comment]
>I'd probably just remove the networking.networkmanager.firewallBackendoption entirely, and keep the autodetection code in NetworkManager itself.

So it should only use nftables then (see other comment)? I'll edit that then

**Comment by flokli** [comment]
SGTM!

File: nixos/doc/manual/release-notes/rl-2311.section.md [hunk 2]
```diff
@@ -217,6 +217,7 @@
   order, or relying on `mkBefore` and `mkAfter`, but may impact users calling
   `mkOrder n` with n ≤ 400.
 
+- `networking.networkmanager.firewallBackend` was removed as nixos is now using iptables-nftables-compat even when using iptables, therefore Networkmanager now uses the nftables backend unchangably.
```
Comment (flokli) [inline nixos/doc/manual/release-notes/rl-2311.section.md:220]:
```suggestion
- `networking.networkmanager.firewallBackend` was removed as nixos is now using iptables-nftables-compat even when using iptables, therefore Networkmanager now uses the nftables backend unconditionally.
```

File: nixos/modules/services/networking/networkmanager.nix [hunk 3]
```diff
@@ -413,6 +402,9 @@ in
       them via the DNS server in your network, or use environment.etc
       to add a file into /etc/NetworkManager/dnsmasq.d reconfiguring hostsdir.
     '')
+    (mkRemovedOptionModule [ "networking" "networkmanager" "firewallBackend" ] ''
+      This option was removed as nixos is now using iptables-nftables-compat even when using iptables, therefore Networkmanager now uses the nftables backend unchangably.
```
Comment (flokli) [inline nixos/modules/services/networking/networkmanager.nix:406]:
```suggestion
      This option was removed as nixos is now using iptables-nftables-compat even when using iptables, therefore Networkmanager now uses the nftables backend unconditionally.
```

**Review by flokli — APPROVED** [review]

**Review by jian-lin — APPROVED** [review]
some nitpicking

**Pull request merged**